src/passes/GUFA.cpp

```diff
@@ -0,0 +1,311 @@
+      // anything about refs.
+      return;
+    }
+
+    // Ok, this is an interesting location that we might optimize. See what the
+    // oracle says is possible there.
+    auto contents = oracle.getContents(ExpressionLocation{curr, 0});
+
+    auto& options = getPassOptions();
+    auto& wasm = *getModule();
+    Builder builder(wasm);
+
+    auto replaceWithUnreachable = [&]() {
+      if (canReplaceWithUnreachable(curr)) {
+        replaceCurrent(getDroppedChildrenAndAppend(
+          curr, wasm, options, builder.makeUnreachable()));
+      } else {
+        // We can't remove this, but we can at least put an unreachable
+        // right after it.
+        replaceCurrent(builder.makeSequence(builder.makeDrop(curr),
+                                            builder.makeUnreachable()));
+      }
+      optimized = true;
+    };
+
+    if (contents.getType() == Type::unreachable) {
+      // This cannot contain any possible value at all. It must be unreachable
+      // code.
+      replaceWithUnreachable();
+      return;
+    }
+
+    // This is reachable. Check if we can emit something optimized for it.
+    // TODO: can we handle more general things here too?
+    if (!contents.canMakeExpression()) {
+      return;
+    }
+
+    if (contents.isNull() && curr->type.isNullable()) {
+      // Null values are all identical, so just fix up the type here if we need
+      // to (the null's type might not fit in this expression, if it passed
+      // through casts).
+      if (!Type::isSubType(contents.getType(), curr->type)) {
+        contents = PossibleContents::literal(
+          Literal::makeNull(curr->type.getHeapType()));
+      }
+
+      // Note that if curr's type is *not* nullable, then the code will trap at
+      // runtime (the null must arrive through a cast that will trap). We handle
+      // that below, so we don't need to think about it here.
+
+      // TODO: would emitting a more specific null be useful when valid?
+    }
+
+    auto* c = contents.makeExpression(wasm);
+
+    // We can only place the constant value here if it has the right type. For
+    // example, a block may return (ref any), that is, not allow a null, but in
+    // practice only a null may flow there if it goes through casts that will
+    // trap at runtime.
+    // TODO: GUFA should eventually do this, but it will require it properly
+    //       filtering content not just on ref.cast as it does now, but also
+    //       ref.as etc. Once it does those we could assert on the type being
+    //       valid here.
+    if (Type::isSubType(c->type, curr->type)) {
+      if (canRemove(curr)) {
+        replaceCurrent(getDroppedChildrenAndAppend(curr, wasm, options, c));
+      } else {
+        // We can't remove this, but we can at least drop it and put the
+        // optimized value right after it.
+        replaceCurrent(builder.makeSequence(builder.makeDrop(curr), c));
```

Comment:
There is some risk here, that is true. Perhaps this should only be done when not optimizing for size? However, propagating the constant to here will potentially open up opportunities to remove other code, so I'm not sure. Will add a comment now but let's keep thinking about it.

Reply:
Maybe we disable this part of code and do a quick check on perf and size, if that's easily doable?

Reply:
Testing now, code size is 1.6% smaller with this optimization. It's harder to test speed (we don't have great wasm GC benchmarks for this kind of thing), but emitting more constants is almost always better. So at least on this one example (j2wasm application) it looks worth doing.

Reply:
Hmm does that mean DCE or Vacuum was eventually able to remove those side-effect-having expressions? I was like, even if we emit an additional constant, if we cannot remove the original expression, that original expression still needs to be run..

Reply:
Yeah, there's no guarantee we can remove it. But sometimes dropping an expression lets us remove parts of it:
```wat
(select
  (call $get-A) ;; returns ref.func $foo
  (call $get-B) ;; also returns ref.func $foo
  (..condition..)
)
=>
(drop
  (select
    (call $get-A) ;; returns ref.func $foo
    (call $get-B) ;; also returns ref.func $foo
    (..condition..)
  )
)
(ref.func $foo)
```
Now that the select is dropped we don't need it, and vacuum will remove it and just leave the calls (and maybe the condition).

(Not sure if that's the main factor here though. Could also be that we don't remove any dropped code, but the constant lets us remove other code...)

Reply:
I'm wondering what if move all meaningful reduction code from vacuum pass to some common util file and reuse routines from this file for vacuum and this pass? At least for those functions that don't require a global analysis of the program, but only subexpressions. This will allow you to avoid code overgrowth and its subsequent removal through a separate pass. It will also allow you to evaluate the benefit of constant propagation ahead of time (at least in local cases).

Reply:
@MaxGraey The vacuum code is already reusable right now, isn't it? We reuse it from here already just by doing `runner.add("vacuum");`. If you mean we just need a subset of vacuum, I'm not sure that's true, we need basically all of it.

Reply:
`runner.add("vacuum")` run on per function level right? But It can't provide a feedback. I mean something like `tryVacuum(transform(expr))` and if `tryVacuum` return false (which mean you can't cleanup code)just don't replace current expression to transformed due to code overgrowth